RemoteImageDecoderAVFProxy leaks video frames through CG image cache

kkinnunen-apple opened the pull request:
#### 614cbf4000576667d412cdbe275f282fd7ad67c5
<pre>
RemoteImageDecoderAVFProxy leaks video frames through CG image cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=261869">https://bugs.webkit.org/show_bug.cgi?id=261869</a>
rdar://115834350

Reviewed by Brent Fulgham.

Currently transferring ImageDecoder frames from GPUP happens for
example when a video is in &lt;img&gt; tag.
Currently the frame is converted to a new bitmap due to NativeImage
transfer shortcomings.

When drawn from, CG might convert the image to a source it can read
more efficiently. Unless the image is marked as transient, the converted
source image will be stored into a global cache. In case of videos,
this cache grows very fast since the videos might be big.

Fix by marking the frame as transient. Later on the conversion-transfer
step is planned to be fixed in more consistent way.

* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::createTransient):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.mm:
(WebCore::ImageDecoderAVFObjC::createFrameImageAtIndex):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::createTransient):
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::createFrameImageAtIndex):

Canonical link: <a href="https://commits.webkit.org/268323@main">https://commits.webkit.org/268323@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f97673f4a80ddc448620c22edc62af24fbeff6b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19338 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19756 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20355 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21241 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18094 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19573 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23025 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19881 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19736 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19555 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19604 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16813 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22105 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16790 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17601 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17852 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17776 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21905 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/18368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15554 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17502 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4624 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21862 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/18198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.mm:
```diff
@@ -579,6 +579,7 @@ SampleFlags flags() const override
 
 PlatformImagePtr ImageDecoderAVFObjC::createFrameImageAtIndex(size_t index, SubsamplingLevel, const DecodingOptions&)
 {
+    WTFReportBacktraceWithPrefix("KKIMMO");
```
Leftover debugging call should be removed. Didn't know about the WithPrefix variant - very useful for filtering logs!

rreno commented on Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
```diff
@@ -31,9 +31,35 @@
+    if (!transientImage)
+        return nullptr;
+    image = nullptr;
+    CGImageSetCachingFlags(transientImage.get(), kCGImageCachingTransient);
```
I want to make sure I have this straight. When we read from the CGImageRef contained in the `image` variable here (later on, somewhere else) we cause CG to put the underlying data into a global cache? And even though we eventually clear it from our caches the retain count is non-zero because of that global cache?

This SPI then tells CG not to cache the underlying data?

kkinnunen-apple replied:
When a CGImage is drawn, the underlying storage is sometimes converted to other formats and cropped and the result is stored to global cache. This is done so that the next time these conversions can be skipped.
Transient flag tells CG that likely there won't be a next time.
This doesn't fix our underlying problems, it's a small step fixing them.

rreno replied:
Ah ok! That makes sense.

brentfulgham commented on Source/WebCore/platform/graphics/NativeImage.h:
```diff
@@ -45,6 +45,8 @@ class NativeImage final : public RenderingResource {
     WTF_MAKE_FAST_ALLOCATED;
 public:
     static WEBCORE_EXPORT RefPtr<NativeImage> create(PlatformImagePtr&&, RenderingResourceIdentifier = RenderingResourceIdentifier::generate());
+    // Creates a NativeImage that will not be used many times. Avoids putting the image to any caches.
```
Slightly clearer phrasing:

`// Creates a NativeImage that will not be used frequently. Avoids storing the image in any caches.`

kkinnunen-apple replied:
Right, but frequently and many times are different.
You can get 100 uses infrequently vs. you can get only 1 use. Then once per 1ms is frequently for somebody, and for other it might be infrequent.. I can rephrase..
.

brentfulgham reviewed (approved):
Looks like a reasonable step forward.

kkinnunen-apple committed:
RemoteImageDecoderAVFProxy leaks video frames through CG image cache
https://bugs.webkit.org/show_bug.cgi?id=261869
rdar://115834350

Reviewed by Brent Fulgham.

Currently transferring ImageDecoder frames from GPUP happens for
example when a video is in <img> tag.
Currently the frame is converted to a new bitmap due to NativeImage
transfer shortcomings.

When drawn from, CG might convert the image to a source it can read
more efficiently. Unless the image is marked as transient, the converted
source image will be stored into a global cache. In case of videos,
this cache grows very fast since the videos might be big.

Fix by marking the frame as transient. Later on the conversion-transfer
step is planned to be fixed in more consistent way.

* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::createTransient):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.mm:
(WebCore::ImageDecoderAVFObjC::createFrameImageAtIndex):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::createTransient):
* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::createFrameImageAtIndex):

Canonical link: https://commits.webkit.org/268323@main
